woocommerce/woocommerce
# Add taxonomy block using SelectTree

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





This creates a "Hierarchical Taxonomy" block, which will be firstly used for WooCommerce Brands plugin, and also replace woocommerce/product-category-field block in the future.

Part of https://github.com/woocommerce/woocommerce/issues/39206 .



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Install the "Custom Post Type UI" plugin by WebDevStudios
2. Set up some custom taxonomies.
    - Make sure to attach it to the Products post type
3. Add code similar to the following to a plugin/mu-plugin file, to add the properties to the REST API response and request:

```php
function YOUR_PREFIX_rest_api_prepare_custom1_to_product( $response, $post ) {
	$post_id = $post->get_id();

	if ( empty( $response->data[ 'custom1' ] ) ) {
		$terms = [];

		foreach ( wp_get_post_terms( $post_id, 'custom-taxonomy' ) as $term ) {
			$terms[] = [
				'id'   => $term->term_id,
				'name' => $term->name,
				'slug' => $term->slug,
			];
		}

		$response->data[ 'custom1' ] = $terms;
	}

	return $response;
}

add_filter( 'woocommerce_rest_prepare_product_object', 'YOUR_PREFIX_rest_api_prepare_custom1_to_product', 10, 2 );

function YOUR_PREFIX_rest_api_add_custom1_to_product( $product, $request, $creating = true ) {
	$product_id = $product->get_id();
	$params     = $request->get_params();
	$custom1s     = isset( $params['custom1'] ) ? $params['custom1'] : array();

	if ( ! empty( $custom1s ) ) {
		if ( $custom1s[0]['id'] ) {
			$custom1s = array_map(
				function ( $custom1 ) {
					return absint( $custom1['id'] );
				},
				$custom1s
			);
		} else {
			$custom1s = array_map( 'absint', $custom1s );
		}
		wp_set_object_terms( $product_id, $custom1s, 'custom-taxonomy' );
	}
}

add_filter( 'woocommerce_rest_insert_product_object', 'YOUR_PREFIX_rest_api_add_custom1_to_product', 10, 3 );
```

4. Add a block for the taxonomy property, using code similar to the following:
```php
$section_fields[] = [
	'woocommerce/taxonomy-field',
	[
		'slug'               => 'custom-taxonomy',
		'property' => 'custom1',
		'label'                                  => __( 'Custom Taxonomy', 'woocommerce' ),
		'createTitle'                        => __( 'Create new custom taxonomy', 'woocommerce' ),
	]
];
```



### Changelog entry




<details>

#### Significance




#### Type




#### Message 

#### Comment 

</details>
